### PR TITLE
10786 setting privateWithin for java protected inner classes

### PR DESCRIPTION
**Pull request by psilospore**
First PR to scala hopefully it's right. This fixes https://github.com/scala/bug/issues/10786

### CODE REVIEW

**Comment by psilospore** [comment]
Here's how I tested it locally:

```
~/Developer/scala/scala/sandbox 10786*
❯ cat Test.java
package pkg;

public class Test { protected static class Inner {} }

~/Developer/scala/scala/sandbox 10786*
❯ cat Use.scala
package pkg

import scala.tools.nsc._
import scala.reflect.runtime.universe._

object Use extends Test.Inner with App {
        println("hi")
        println(typeOf[_root_.pkg.Test].companion.decl(TypeName("Inner")).privateWithin)

}


~/Developer/scala/scala/sandbox 10786*
❯ javac -d . Test.java

~/Developer/scala/scala/sandbox 10786*
❯ ../build/pack/bin/scalac -cp . Use.scala
Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=512m; support was removed in 8.0

~/Developer/scala/scala/sandbox 10786*
❯ ../build/pack/bin/scala pkg.Use
Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=512m; support was removed in 8.0
hi
package pkg

~/Developer/scala/scala/sandbox 10786*
```

I see that in the [scala hacker guide](https://www.scala-lang.org/contribute/hacker-guide.html) it mentions: "Code which should compile successfully, but doesn’t need to be executed, needs to go into the “pos” directory."

Would I just place those files in pos then (minus the print statements)?

File: src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala [hunk 1]
```diff
@@ -1143,6 +1143,9 @@ abstract class ClassfileParser {
         (newStub(name.toTypeName), newStub(name.toTermName))
       } else {
         val cls = owner.newClass(name.toTypeName, NoPosition, sflags)
+        if (cls.isProtected) {
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala]: Looks like the right spot to me. The other cases seem to use `propagatePackageBoundary`. I'm not super familiar with this part. Assigning lrytz for review :-)

**Comment by adriaanm** [comment]
Welcome, @psilospore, and congrats already on your first scalac bug fix 👏 🍰

File: src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala [hunk 2]
```diff
@@ -1143,6 +1143,9 @@ abstract class ClassfileParser {
...
+          cls.privateWithin = cls.enclosingPackage
+        }
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala]: This looks good, I think it needs to be done for the `mod` below too. A single check `if (jflags.isProtected)` should be enough.

**Review by lrytz — CHANGES_REQUESTED** [review]
It took me a looong while to understand what's going on here, but in the end your change seems to be the correct fix.

Here's my test case:

A.java
```java
package a.b;

public class A {
  static class N1 {
    public int x = 1;
    public static int y = 1;
  }
  static protected class N2 {
    public int x = 1;
    public static int y = 1;
  }
}
```

Test.sclaa
```scala
package a.b {
  class C {
    class T1 extends A.N1
    class T2 extends A.N2
    def test(): Unit = {
      val n1 = new A.N1
      n1.x
      A.N1.y
      val n2 = new A.N2
      n2.x
      A.N2.y
    }

    class I extends A {
      class T1 extends A.N1
      class T2 extends A.N2
      def test(): Unit = {
        val n1 = new A.N1
        n1.x
        A.N1.y
        val n2 = new A.N2
        n2.x
        A.N2.y
      }
    }
  }
}
```

I was wondering why the bug only affects `N2`, but not `N1`. What's going on:
  - The classfile parser for `A` creates symbols for all nested classes (`enterOwnInnerClasses`)
  - When those inner symbols complete, the classfile parser runs on `A$N1` / `A$N2` etc
    - There is an invocation of `propagatePackageBoundary`, which sets the `privateWithin` for `N1`, because the classfile `A$N1` has no access modifier
    - However, this does not happen for `A$N2`, because that classfile is marked `public` in bytecode (!). The JVM access rules require making the class public (this will probably change on Java 11 with nestmates?)

Could you
  - Add a comment to your fix? Something like "need to set `privateWithin` here because the classfile of a nested protected class is public in bytecode, so `propagatePackageBoundary` will not set it when the symbols are completed"
  - Extend the test case to test access to a static member within the nested class

File: src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala [hunk 3]
```diff
@@ -1143,16 +1143,22 @@ abstract class ClassfileParser {
...
         mod.moduleClass setInfo loaders.moduleClassLoader
         cls.associatedFile = file
         mod.moduleClass.associatedFile = file
+
+        /**
+          * need to set privateWithin here because the classfile of a nested protected class is public in bytecode,
+          * so propagatePackageBoundary will not set it when the symbols are completed
+          */
+        if (jflags.isProtected) {
+          cls.privateWithin = cls.enclosingPackage
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala:1159]: I think we also need to set `mod.privateWithin = ...`, see the test case in my comment https://github.com/scala/scala/pull/7323#pullrequestreview-163307986. Add an instance and a static member to the two nested java classes `StaticClass` and `ProtectedStaticClass`, and access these members from Scala code.

**Review by lrytz — CHANGES_REQUESTED** [review]

**Review by lrytz — COMMENTED** [review]
Great, thank you! This looks good now. Could you squash all commits into a single one and write a summary for the commit message?

**Comment by psilospore** [comment]
No problem!

Thanks, @lrytz, @adriaanm, and @hrhino I appreciate all the help!

**Review by lrytz — APPROVED** [review]

**Comment by adriaanm** [comment]
Thanks, @psilospore!

**Comment by hrhino** [comment]
congrats on your first PR, @psilospore!

**Pull request merged**